### PR TITLE
Toast alerts

### PR DESCRIPTION
**Pull request by ianmclean2011**
Just added a minor fix for when a toast is closed by removing the open attribute, as well as aria-alerts.

### CODE REVIEW

File: d2l-alert-toast.html [hunk 1]
```diff
@@ -150,12 +151,19 @@
...
+						this.classList.remove('opened');
+						this.removeAttribute('role');
+					}
 				}
 			},
 
 			_close: function() {
 				this.classList.add('closed');
 				this.classList.remove('opened');
+				this.removeAttribute('role');
```
Comment (dlockhart) [inline d2l-alert-toast.html]: This line seems redundant? When you set `this.open = false;` on the next line, won't `_changeOpen()` get called and it'll remove the attribute there?

Comment (ianmclean2011) [reply d2l-alert-toast.html]: Yeah, that looks right. It looks like line 165 and 166 are now redundant, so I'll remove them, and test that out a bit. Thanks!

File: d2l-alert-toast.html [hunk 2]
```diff
@@ -153,7 +153,9 @@
 					}
 				} else {
 					if (this.classList.contains('opened')) {
-						this.classList.add('closed');
+						if (!this.classList.contains('closed')) {
+							this.classList.add('closed');
+						}
```
Comment (dlockhart) [inline d2l-alert-toast.html]: The `classList` collection's `add` and `remove` are pretty smart -- they're only going to add stuff if it's not there already (it won't get added twice). So the `if`-statement is not needed.

Comment (ianmclean2011) [reply d2l-alert-toast.html]: Awesome ... I'll get rid of it ... thanks!

Comment (ianmclean2011) [reply d2l-alert-toast.html]: From all through the code I wrote.

File: d2l-alert-toast.html [hunk 3]
```diff
@@ -128,13 +128,10 @@
...
 					requestAnimationFrame(function() {
-						if (this.classList.contains('closed')) {
-							this.classList.remove('closed');
-						}
+						this.classList.remove('closed');
 						requestAnimationFrame(function() {
-							if (!this.classList.contains('opened')) {
-								this.classList.add('opened');
-							}
+							this.classList.add('opened');
```
Comment (dbatiste) [inline d2l-alert-toast.html]: I think it would be good to avoid class names that are generic like this, especially at the host level of the component.  There is risk of something else defining such style and inadvertently affecting this component.  If using a class, I think it would be better to use `d2l-alert-toast-opened`.  Alternatively, an attribute selector could be used.

The other thing about this is that the classList is technically part of the public API, so there ends up being two public representations of the opened/closed state - the attribute/property, and the class.  If a class is indeed needed, my preference is to keep it internal (not on the host).

Comment (ianmclean2011) [reply d2l-alert-toast.html]: I agree about the generic classnames, and I'm happy to switch to `d2l-alert-toast-opened` or something like that. I know that we could use attribute selectors, and perhaps that's a better route, and something I can look at in the near future. Initially, I was hoping to just use the `open` attribute, however, the animation and setting `display:none` complicated things, and classes were just more familiar to me.

I don't really understand your concern in the second paragraph. Perhaps that goes away if I switch the class to be less generic? 

If it's agreeable, I'd like to switch to a non-generic class for now, and look at ways to make it better in the future. Does that sound good?

Comment (dbatiste) [reply d2l-alert-toast.html]: Re the 2nd paragraph, there are really two issues.  

The first is that the classList is technically part of the public API, so using it for internal implementation is kinda risky because a consumer could assign a new class value, unintentionally removing the class, resulting in a weird bugged-out state.

I know in some components we might apply a class to the host during an animation, and then we may remove it at some point a few frames later.  It's not ideal either because it's technically it's an internal detail exposed in a public way, and subject to be manipulated by the consuming code.  I am sure there are other components that sprout classes like this, and we can probably get away with it if the class names aren't too generic, but it still isn't great.

The second issue is that it results in two sources truth (the attribute, and the class), which requires code to keep them in sync.  If they get out of sync for some reason, the state of the component gets messed up.
```
<d2l-alert-toast open class="... d2l-alert-toast-opened ...">
```

Comment (dlockhart) [reply d2l-alert-toast.html]: Dave just means that if you inspect the element in dev tools, you'll see `<d2l-alert-toast open class="opened">`, so it's "public" and other developers will see it and might do crazy stuff with it (like adding/removing it themselves or writing CSS to target it). If instead you apply it to an element inside, it'll be hidden in shadow DOM.

Comment (dlockhart) [reply d2l-alert-toast.html]: If the class is just used for animations, I like the idea of calling it something like `d2l-alert-toast-opening` or `-closing` and then removing it entirely when the animation is finished.

Comment (ianmclean2011) [reply d2l-alert-toast.html]: Ok ... cool ... I think I understand the general concern now. Let me see what I can do about just using the open attribute to represent being open or closed (as the only public facing thing), and handle animation stuff underneath (or with temporary non-generic classes).

Comment (dlockhart) [reply d2l-alert-toast.html]: Yeah you're right, an attribute with 2 states isn't enough to do the animation properly. But a multi-state attribute will make it hard for consumers to interact with it, since they should only need to add or remove the `open` attribute -- that's all they care about.

Personally I'd probably explore the internal `div` approach. You'll need to move the `display: none` to that `div`, but could then combine a parent selector for `:host([open]) d2l-alert-toast-container` with a CSS class on the container to trigger its `display: block` and animations.

**Comment by ianmclean2011** [comment]
This is working a lot better! A bit cleaner, and transition related classes are internal. Let me know what you think!

Also, as a side note form testing, for some reason in Edge, using the mainline demo, the message highlight for alerts (the color indicator on the side of the alert) is picking up RTL styling when it shouldn't be, and is backwards. Not sure if you've seen this in other projects, and might have an idea why. It looks fine in the 1.x Edge demo, and looks fine in all other browsers and demos.

File: d2l-alert-toast.html [hunk 4]
```diff
@@ -9,11 +9,16 @@
 
 <dom-module id="d2l-alert-toast">
 	<template strip-whitespace>
-		<style>
-			:host {
+		<style>
```
Comment (dlockhart) [inline d2l-alert-toast.html]: By default, `:host` is going to have a `display: inline`. I'd recommend adding `:host { display: block; }` to be explicit about it.

File: d2l-alert-toast.html [hunk 5]
```diff
@@ -9,11 +9,16 @@
...
 <dom-module id="d2l-alert-toast">
 	<template strip-whitespace>
-		<style>
-			:host {
+		<style>
+			:host .d2l-alert-toast-container {
+				display: none;
+			}
+
+			:host([open]) .d2l-alert-toast-container{
```
Comment (dlockhart) [inline d2l-alert-toast.html]: Nit: missing space

File: d2l-alert-toast.html [hunk 6]
```diff
@@ -27,22 +32,20 @@
...
 				width: 100%;
 				z-index: 10000; /*This was based on one of the flash messages in the LMS ... if need be, adjust to be more appropriate*/
-			}
-
-			:host(.closed) {
-				display:none;
 			}
 
-			:host(.opened) {
+			:host([open]) .d2l-alert-toast-container.opened {
```
Comment (dlockhart) [inline d2l-alert-toast.html]: I'd still recommend using a CSS class name like `d2l-alert-toast-container-opened` since someone could easily define a `opened` CSS class in the global namespace that could bleed into this.

**Comment by dlockhart** [comment]
For the Edge issue, can you tell if it's a Polymer 2 (mainline) vs. Polymer 1 issue, or a shadow vs. shady DOM issue? Polymer 2 has shadow DOM enabled by default vs. Polymer 1 won't. But you can enable shadow DOM in Polymer 1 by passing `?dom=shadow` on the query-string.

Another way to debug would be to use the dev tools in Edge (sorry) and inspect the actual CSS rules that are getting applied to the elements. Hopefully that'll give a hint as to why the RTL styles are triggering when they shouldn't be.

File: d2l-alert-toast.html (hunk 6 above)
Comment (ianmclean2011) [reply d2l-alert-toast.html]: Sure thing ... I've changed the opened class to d2l-alert-toast-container-opened.

Just so I can confirm my understanding ... is the concern of bleeding into the component when using shadydom, or does that concern also exist for shadowdom? Just curious, as I thought shadowdom was supposed to prevent that.

Comment (dbatiste) [reply d2l-alert-toast.html]: While shadow-DOM will isolate it's nodes from CSS bleeding/cascading from parents, if the class is placed on the host element then it will be subject for targeting within that DOM context (which may or may not be in another shadow-DOM).  So the practice of qualifying CSS class names is still worthwhile.  That being said, I think it's not a great practice for the component to use the class of the host for the other reasons mentioned.
